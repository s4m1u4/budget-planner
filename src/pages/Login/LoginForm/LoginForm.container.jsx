import { inject, observer } from "mobx-react";
import LoginForm from "./LoginForm";

const LoginFormContainer = inject(
  ({
    rootStore: {
      userStore: { setIsAuth },
    },
  }) => ({ setIsAuth })
)(observer(LoginForm));

export default LoginFormContainer;
