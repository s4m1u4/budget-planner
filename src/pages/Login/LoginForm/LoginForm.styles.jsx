export const box = {
  margin: "0 auto",
  width: "100%",
  maxWidth: "400px",
};

export const title = {
  textAlign: "center",
  margin: "0 0 10px 0",
};

export const form = {
  display: "flex",
  flexDirection: "column",
  rowGap: "15px",
  margin: "0 0 15px 0",
  padding: "20px",
  border: "1px solid rgb(208, 215, 222)",
  borderRadius: "6px",
};

export const info = {
  padding: "15px 20px",
  border: "1px solid rgb(208, 215, 222)",
  borderRadius: "6px",
};
