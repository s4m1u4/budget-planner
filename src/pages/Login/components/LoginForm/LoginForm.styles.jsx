export const title = {
  textAlign: "center",
};

export const boxForm = {
  width: "100%",
  maxWidth: "400px",
};

export const form = {
  display: "flex",
  flexDirection: "column",
  rowGap: "15px",
  margin: "15px 0",
  padding: "20px",
  border: "1px solid rgb(208, 215, 222)",
  borderRadius: "6px",
};

export const boxInfo = {
  padding: "15px 20px",
  border: "1px solid rgb(208, 215, 222)",
  borderRadius: "6px",
};
