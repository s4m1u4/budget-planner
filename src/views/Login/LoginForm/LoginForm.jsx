import React from "react";
import * as yup from "yup";
import { useFormik } from "formik";
import { Button, TextField, Typography } from "@mui/material";
import { Link } from "react-router-dom";
import styled from "@emotion/styled";

const LoginSchema = yup.object({
  email: yup
    .string("Enter your email")
    .email("Enter a valid email")
    .required("Email is required"),
  password: yup
    .string("Enter your password")
    .min(8, "Password should be of minimum 8 characters length")
    .required("Password is required"),
});

const LoginForm = () => {
  const Box = styled.div({
    margin: "0 auto",
    padding: "20px 0",
    width: "100%",
    maxWidth: "350px",
  });

  const Form = styled.form({
    display: "flex",
    flexDirection: "column",
    rowGap: "15px",
    margin: "0 0 15px 0",
    padding: "20px",
    border: "1px solid rgb(208, 215, 222)",
    borderRadius: "6px",
  });

  const Wrapper = styled.div({
    padding: "15px 20px",
    border: "1px solid rgb(208, 215, 222)",
    borderRadius: "6px",
  });

  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    validationSchema: LoginSchema,
    onSubmit: (values) => {
      alert(JSON.stringify(values, null, 2));
      formik.resetForm();
    },
  });

  return (
    <Box>
      <Typography
        variant="h6"
        component="h1"
        sx={{ textAlign: "center", margin: "0 0 10px 0" }}
      >
        Sign in
      </Typography>
      <Form onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          name="email"
          type="text"
          size="small"
          label="Email"
          value={formik.values.email}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.email && !!formik.errors.email}
          helperText={formik.touched.email && formik.errors.email}
        />
        <TextField
          fullWidth
          name="password"
          type="password"
          size="small"
          label="Password"
          value={formik.values.password}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.password && !!formik.errors.password}
          helperText={formik.touched.password && formik.errors.password}
        />
        <Button variant="contained" color="success" type="submit">
          Login
        </Button>
      </Form>
      <Wrapper>
        <Typography variant="body2" component="p">
          Don't have an account yet?{" "}
          <Link to="/signup">Create an account.</Link>
        </Typography>
      </Wrapper>
    </Box>
  );
};

export default LoginForm;
