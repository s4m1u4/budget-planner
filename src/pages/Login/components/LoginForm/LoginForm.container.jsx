import { inject, observer } from "mobx-react";
import LoginForm from "./LoginForm";

const LoginFormContainer = inject(
  ({
    rootStore: {
      userStore: { userAuthentication },
    },
  }) => ({ userAuthentication })
)(observer(LoginForm));

export default LoginFormContainer;
