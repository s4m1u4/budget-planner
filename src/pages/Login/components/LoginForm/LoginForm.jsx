import React from "react";
import { useFormik } from "formik";
import { Link, useNavigate } from "react-router-dom";
import { Box, Button, TextField, Typography } from "@mui/material";
import { LoginFormSchema } from "./LoginFormSchema";

import { boxForm, boxInfo, form, title } from "./LoginForm.styles";

export const LoginForm = ({ userAuthentication }) => {
  const navigate = useNavigate();

  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    validationSchema: LoginFormSchema,
    onSubmit: async (values) => {
      await userAuthentication(values);
      navigate("/");
      formik.resetForm();
    },
  });

  return (
    <Box sx={boxForm}>
      <Typography variant="h6" component="h1" sx={title}>
        Sign in
      </Typography>
      <Box sx={form} component="form" onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          name="email"
          type="text"
          size="small"
          label="Email"
          value={formik.values.email}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.email && !!formik.errors.email}
          helperText={formik.touched.email && formik.errors.email}
        />
        <TextField
          fullWidth
          name="password"
          type="password"
          size="small"
          label="Password"
          value={formik.values.password}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.password && !!formik.errors.password}
          helperText={formik.touched.password && formik.errors.password}
        />
        <Button variant="contained" color="success" type="submit">
          Sign in
        </Button>
      </Box>
      <Box sx={boxInfo}>
        <Typography variant="body2" component="p">
          Don't have an account yet? <Link to="/signup">Create an account</Link>
        </Typography>
      </Box>
    </Box>
  );
};
