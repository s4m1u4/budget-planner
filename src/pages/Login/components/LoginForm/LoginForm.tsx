import React, { ChangeEvent, FC, useEffect, useState } from "react";
import { useFormik } from "formik";
import { Link, useNavigate } from "react-router-dom";
import { Alert, Box, Button, TextField, Typography } from "@mui/material";
import { LoginFormSchema } from "./LoginFormSchema";
import { LoginProps } from "../../types";
import { IUserAuthenticationData } from "../../../../types";

import { boxForm, boxInfo, form, title } from "./LoginForm.styles";

export const LoginForm: FC<LoginProps> = ({ userAuthentication, onSubmit }) => {
  const navigate = useNavigate();
  const [errorMessage, setErrorMessage] = useState<string>("");

  useEffect(() => {
    return () => {
      setErrorMessage("");
    };
  }, []);

  const handleSubmit = async (values: IUserAuthenticationData) => {
    const response: string | void = await userAuthentication(values);
    response && setErrorMessage(response);
    !response && navigate("/dashboard");
    !response && formik.resetForm();
  };

  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    validationSchema: LoginFormSchema,
    onSubmit: async (values: IUserAuthenticationData) => {
      await handleSubmit(values);
      await onSubmit(values);
    },
  });

  const formikHandleChange = (event: ChangeEvent): void => {
    formik.handleChange(event);
    setErrorMessage("");
  };

  return (
    <Box sx={boxForm}>
      <Typography variant="h6" component="h1" sx={title}>
        Sign in
      </Typography>
      <Box sx={form} component="form" onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          id="email"
          name="email"
          type="text"
          size="small"
          label="Email"
          value={formik.values.email}
          onBlur={formik.handleBlur}
          onChange={formikHandleChange}
          error={formik.touched.email && !!formik.errors.email}
          helperText={formik.touched.email && formik.errors.email}
        />
        <TextField
          fullWidth
          id="password"
          name="password"
          type="password"
          size="small"
          label="Password"
          value={formik.values.password}
          onBlur={formik.handleBlur}
          onChange={formikHandleChange}
          error={formik.touched.password && !!formik.errors.password}
          helperText={formik.touched.password && formik.errors.password}
        />
        <Button variant="contained" color="success" type="submit">
          Sign in
        </Button>
      </Box>
      {errorMessage && (
        <Alert variant="filled" severity="error" sx={{ marginBottom: "15px" }}>
          {errorMessage}
        </Alert>
      )}
      <Box sx={boxInfo}>
        <Typography variant="body2" component="p">
          Don't have an account yet? <Link to="/signup">Create an account</Link>
        </Typography>
      </Box>
    </Box>
  );
};
