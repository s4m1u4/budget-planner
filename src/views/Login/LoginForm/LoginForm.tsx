import React, { FC } from "react";
import { useFormik } from "formik";
import userStore from "../../../store/UserStore";
import { Link } from "react-router-dom";
import { Box, Button, TextField, Typography } from "@mui/material";
import { LoginSchema } from "./LoginForm.schema";
import { box, form, info, title } from "./LoginForm.styles";

const LoginForm: FC = () => {
  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    validationSchema: LoginSchema,
    onSubmit: (values) => {
      userStore.setIsAuth();
      formik.resetForm();
    },
  });

  return (
    <Box sx={box}>
      <Typography variant="h6" component="h1" sx={title}>
        Sign in
      </Typography>
      <Box sx={form} component="form" onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          name="email"
          type="text"
          size="small"
          label="Email"
          value={formik.values.email}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.email && !!formik.errors.email}
          helperText={formik.touched.email && formik.errors.email}
        />
        <TextField
          fullWidth
          name="password"
          type="password"
          size="small"
          label="Password"
          value={formik.values.password}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.password && !!formik.errors.password}
          helperText={formik.touched.password && formik.errors.password}
        />
        <Button variant="contained" color="success" type="submit">
          Sign in
        </Button>
      </Box>
      <Box sx={info}>
        <Typography variant="body2" component="p">
          Don't have an account yet? <Link to="/signup">Create an account</Link>
        </Typography>
      </Box>
    </Box>
  );
};

export default LoginForm;
