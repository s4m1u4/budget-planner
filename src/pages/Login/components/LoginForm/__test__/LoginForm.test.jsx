import { LoginForm } from "../LoginForm";
import { BrowserRouter } from "react-router-dom";
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const renderWithRouter = () => {
  return render(
    <BrowserRouter>
      <LoginForm onSubmit={onSubmit} userAuthentication={userAuthentication} />
    </BrowserRouter>
  );
};

const renderWithRouterAsync = async () => {
  await act(async () => {
    await render(
      <BrowserRouter>
        <LoginForm
          onSubmit={onSubmit}
          userAuthentication={userAuthentication}
        />
      </BrowserRouter>
    );
  });
};

const onSubmit = jest.fn();
const userAuthentication = jest.fn();

describe("Log in form component", () => {
  it("snapshot matching", async () => {
    const { baseElement } = renderWithRouter();
    expect(baseElement).toMatchSnapshot();
  });

  beforeEach(async () => {
    await renderWithRouterAsync();
  });

  describe("should render", () => {
    it("an input with name='email'", () => {
      expect(
        screen.getByRole("textbox", { name: /email/i })
      ).toBeInTheDocument();
    });

    it("an input with name='password'", () => {
      expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
    });

    it("a button 'Sing in'", () => {
      expect(
        screen.getByRole("button", { name: /sign in/i })
      ).toBeInTheDocument();
    });

    it("error when input values are empty", async () => {
      expect(screen.queryByText(/email is required/i)).toBeNull();
      expect(screen.queryByText(/password is required/i)).toBeNull();

      userEvent.type(screen.getByRole("textbox", { name: /email/i }), "");
      userEvent.type(screen.getByLabelText(/password/i), "");

      userEvent.click(screen.getByRole("button", { name: /sign in/i }));

      await renderWithRouterAsync();

      expect(screen.getByText(/email is required/i)).toBeInTheDocument();
      expect(screen.getByText(/password is required/i)).toBeInTheDocument();
    });
  });

  it("successful form submission", async () => {
    userEvent.type(
      screen.getByRole("textbox", { name: /email/i }),
      "totalgol2015@gmail.com"
    );
    userEvent.type(screen.getByLabelText(/password/i), "v4d1kRed2k");

    userEvent.click(screen.getByRole("button", { name: /sign in/i }));

    await waitFor(() =>
      expect(onSubmit).toHaveBeenCalledWith({
        email: "totalgol2015@gmail.com",
        password: "v4d1kRed2k",
      })
    );

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
