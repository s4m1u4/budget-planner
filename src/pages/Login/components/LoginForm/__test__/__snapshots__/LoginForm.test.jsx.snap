// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log in form component snapshot matching 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-o8u0d7"
    >
      <h1
        class="MuiTypography-root MuiTypography-h6 css-1kgxqm0-MuiTypography-root"
      >
        Sign in
      </h1>
      <form
        class="MuiBox-root css-hbh1ay"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="email"
            id="email-label"
          >
            Email
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="email"
              name="email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            Password
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ke5b6m-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Sign in
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
      
      <div
        class="MuiBox-root css-10kxgn8"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          Don't have an account yet? 
          <a
            href="/signup"
          >
            Create an account
          </a>
        </p>
      </div>
    </div>
  </div>
  <div>
    <div
      class="MuiBox-root css-o8u0d7"
    >
      <h1
        class="MuiTypography-root MuiTypography-h6 css-1kgxqm0-MuiTypography-root"
      >
        Sign in
      </h1>
      <form
        class="MuiBox-root css-hbh1ay"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="email"
            id="email-label"
          >
            Email
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="email"
              name="email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            Password
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ke5b6m-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Sign in
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
      
      <div
        class="MuiBox-root css-10kxgn8"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          Don't have an account yet? 
          <a
            href="/signup"
          >
            Create an account
          </a>
        </p>
      </div>
    </div>
  </div>
</body>
`;
